doc/src/guides/custom-functions.md

```diff
@@ -0,0 +1,1773 @@
+
+```
+>>> from sympy import expand_trig
+>>> y = symbols('y')
+>>> class versin(Function):
+...    def _eval_expand_trig(self, **hints):
+...        x = self.args[0]
+...        return expand_trig(1 - cos(x))
+>>> versin(x + y).expand(trig=True)
+sin(x)*sin(y) - cos(x)*cos(y) + 1
+```
+
+A more sophisticated implementation might attempt to rewrite the result of
+`expand_trig(1 - cos(x))` back into `versin` functions. This is left as an
+exercise for the reader.
+
+(custom-functions-differentiation)=
+### Differentiation
+
+To define differentiation via {func}`~.diff`, define a method `fdiff(self,
+argindex)`. `fdiff()` should return the derivative of the function, without
+considering the chain rule, with respect to the `argindex`-th variable.
+`argindex` is indexed starting at `1`.
+
+That is, `f(x1, ..., xi, ..., xn).fdiff(i)` should return $\frac{d}{d x_i}
+f(x_1, \ldots, x_i, \ldots, x_n)$, where $x_k$ are independent of one another.
+`diff()` will automatically apply the chain rule using the result of
+`fdiff()`. User code should use `diff()` and not call `fdiff()` directly.
+
+```{note}
+
+`Function` subclasses should define differentiation using `fdiff()`. Subclasses
+of {class}`~.Expr` that aren't `Function` subclasses will need to define
+`_eval_derivative()` instead. It is not recommended to redefine
+`_eval_derivative()` on a `Function` subclass.
+
+```
+
+For our [$\operatorname{versin}$ example
+function](custom-functions-versine-definition), the derivative is $\sin(x)$.
+
+```py
+>>> class versin(Function):
+...     def fdiff(self, argindex=1):
+...         return sin(self.args[0])
+```
+
+(custom-functions-differentiation-examples)=
+```py
+>>> versin(x).diff(x)
+sin(x)
+>>> versin(x**2).diff(x)
+2*x*sin(x**2)
+>>> versin(x + y).diff(x)
+sin(x + y)
+```
+
+As an example of a function that has multiple arguments, consider the [fused
+multiply-add (FMA) example](custom-functions-fma-definition) defined above
+($\operatorname{FMA}(x, y, z) = xy + z$).
+
+We have
+
+$$\frac{d}{dx} \operatorname{FMA}(x, y, z) = y,$$
+$$\frac{d}{dy} \operatorname{FMA}(x, y, z) = x,$$
+$$\frac{d}{dz} \operatorname{FMA}(x, y, z) = 1.$$
+
+So the `fdiff()` method for `FMA` would look like this:
+
+```py
+>>> from sympy import Number, symbols
+>>> x, y, z = symbols('x y z')
+>>> class FMA(Function):
+...     """
+...     FMA(x, y, z) = x*y + z
+...     """
+...     def fdiff(self, argindex):
+...         x, y, z = self.args
+...         if argindex == 1:
+...             return y
+...         elif argindex == 2:
+...             return x
+...         elif argindex == 3:
+...             return 1
+```
+
+```py
+>>> FMA(x, y, z).diff(x)
+y
+>>> FMA(x, y, z).diff(y)
+x
+>>> FMA(x, y, z).diff(z)
+1
+>>> FMA(x**2, x + 1, y).diff(x)
+x**2 + 2*x*(x + 1)
+```
+
+To leave a derivative unevaluated, raise
+`sympy.core.function.ArgumentIndexError(self, argindex)`. This is the default
+behavior if `fdiff()` is not defined. Here is an example function $f(x, y)$ that
```

Comment:
Does returning `None` not work?

Reply:
https://github.com/sympy/sympy/issues/23490